### PR TITLE
add keyboard support for duckboard

### PR DESCRIPTION
**Pull request by doodboard**
Adding duckboard support to QMK

## Description

Adding support for Duckboard to QMK. 
Added new folder "duckboard" under keyboard. 
Added config.h / kb.c / kb.h / Makefile / rules.mk / Keymap.c to relevant directory.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x ] Keyboard (addition or update)
- [x ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/doodboard/duckboard/kb.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	K40, K41, K42, K43, K44  \
+) { \
+	{ KC_NO, K01,   K02,   K03,   K04 }, \
+	{ KC_NO, K11,   K12,   K13,   KC_NO }, \
+	{ KC_NO, K21,   K22,   K23,   K24 }, \
+	{ KC_NO, K31,   K32,   K33,   KC_NO }, \
+	{ K40,   K41,   K42,   K43,   K44 }  \
+}
+
+#endif
```
Comment (drashna) [inline keyboards/doodboard/duckboard/kb.h:36]:
```suggestion
#define LAYOUT( \
         K01, K02, K03, K04, \
         K11, K12, K13,      \
         K21, K22, K23, K24, \
         K31, K32, K33,      \
    K40, K41, K42, K43, K44  \
) { \
    { KC_NO, K01,   K02,   K03,   K04 },   \
    { KC_NO, K11,   K12,   K13,   KC_NO }, \
    { KC_NO, K21,   K22,   K23,   K24 },   \
    { KC_NO, K31,   K32,   K33,   KC_NO }, \
    { K40,   K41,   K42,   K43,   K44 }    \
}

```

File: keyboards/doodboard/duckboard/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		TG(2), KC_TRNS, KC_INS, KC_DEL, KC_TRNS),
+
+	KEYMAP(
+		KC_TRNS, RGB_TOG, RGB_MOD, KC_TRNS,
+		RGB_HUI, RGB_SAI, RGB_VAI,
+		RGB_HUD, RGB_SAD, RGB_VAD, KC_TRNS,
+		KC_TRNS, KC_TRNS, KC_TRNS,
+		TG(2), RESET, KC_TRNS, KC_TRNS, KC_TRNS),
+
+};
```
Comment (drashna) [inline keyboards/doodboard/duckboard/keymaps/default/keymap.c:42]:
```suggestion
#include QMK_KEYBOARD_H

const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [0] = LAYOUT(
                 TG(1),   KC_PSLS, KC_PAST, KC_PMNS,
                 KC_P7,   KC_P8,   KC_P9,
                 KC_P4,   KC_P5,   KC_P6,   KC_PPLS,
                 KC_P1,   KC_P2,   KC_P3,
        KC_BSPC, KC_MUTE, KC_P0,   KC_PDOT, KC_PENT),

	[1] = LAYOUT(
		         TG(1),   KC_TRNS, KC_TRNS, KC_TRNS,
		         KC_HOME, KC_UP,   KC_PGUP,
		         KC_LEFT, KC_TRNS, KC_RGHT, KC_TRNS,
		         KC_END,  KC_DOWN, KC_PGDN,
		TG(2),   KC_TRNS, KC_INS,  KC_DEL,  KC_TRNS),

	[2] = LAYOUT(
		         KC_TRNS, RGB_TOG, RGB_MOD, KC_TRNS,
		         RGB_HUI, RGB_SAI, RGB_VAI,
		         RGB_HUD, RGB_SAD, RGB_VAD, KC_TRNS,
		         KC_TRNS, KC_TRNS, KC_TRNS,
		TG(2),   RESET,   KC_TRNS, KC_TRNS, KC_TRNS),

};
```

File: keyboards/doodboard/duckboard/rules.mk [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ARCH = AVR8
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+SRC += ssd1306.c
+
+# Boot Section Size in *bytes*
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
Comment (drashna) [inline keyboards/doodboard/duckboard/rules.mk:14]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by doodboard** [comment]
Sorry for the mistakes. This is my first time contributing and there is a lot to learn, clearly. Thank you in advance for the time and effort!

File: keyboards/doodboard/duckboard/config.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x6060
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    doodboard
+#define PRODUCT         duckboard
+#define DESCRIPTION     18 key numpad with rotary encoder and OLED support
```
Comment (fauxpark) [inline keyboards/doodboard/duckboard/config.h:27]:
```suggestion
```

File: keyboards/doodboard/duckboard/config.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x6060
```
Comment (fauxpark) [inline keyboards/doodboard/duckboard/config.h:23]:
```suggestion
#define PRODUCT_ID      0xFF44
```
```
echo -n "duckboard" | shasum | cut -c1-4
```

File: keyboards/doodboard/duckboard/rules.mk [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# MCU name
+MCU = atmega32u4
+BOOTLOADER = caterina
```
Comment (fauxpark) [inline keyboards/doodboard/duckboard/rules.mk:3]:
```suggestion
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = caterina
```

File: keyboards/doodboard/duckboard/rules.mk [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+CONSOLE_ENABLE ?= yes	# Console for debug(+400)
+COMMAND_ENABLE ?= no    # Commands for debug and configuration
+SLEEP_LED_ENABLE ?= no  # Breathing sleep LED during USB suspend
+NKRO_ENABLE ?= yes		# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE ?= no  # Enable keyboard backlight functionality
+AUDIO_ENABLE ?= no
+RGBLIGHT_ENABLE ?= yes
+
+ENCODER_ENABLE = yes
+OLED_DRIVER_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/doodboard/duckboard/rules.mk:20]:
```suggestion
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output

ENCODER_ENABLE = yes
OLED_DRIVER_ENABLE = yes
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
This board is also missing an info.json.

**Comment by doodboard** [comment]
comments applied, and I added info.json!

**Comment by doodboard** [comment]
the firmware compiles fine on my end, I'm not sure why it's failing the checks

File: keyboards/doodboard/duckboard/duckboard.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define KB_H
```
Comment (fauxpark) [inline keyboards/doodboard/duckboard/duckboard.h:17]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/doodboard/duckboard/info.json [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+      "LAYOUT_default": {
+        "layout": [                 {"x":1, "y":0}, {"x":2, "y":0}, {"x":3, "y":0}, {"x":4, "y":0},
+                                    {"x":1, "y":1}, {"x":2, "y":1}, {"x":3, "y":1},
+                                    {"x":1, "y":2}, {"x":2, "y":2}, {"x":3, "y":2}, {"x":4, "y":2},
+                                    {"x":1, "y":3}, {"x":2, "y":3}, {"x":3, "y":3},
+                    {"x":0, "y":4}, {"x":1, "y":4}, {"x":2, "y":4}, {"x":3, "y":4}, {"x":4, "y":4},
+                  ]
+      }
+    }
+  }
```
Comment (fauxpark) [inline keyboards/doodboard/duckboard/info.json:17]:
```suggestion
{
    "keyboard_name": "duckboard",
    "url": "",
    "maintainer": "doodboard",
    "width": 5,
    "height": 5,
    "layouts": {
        "LAYOUT_default": {
            "layout": [
                {"x": 1, "y": 0},
                {"x": 2, "y": 0},
                {"x": 3, "y": 0},
                {"x": 4, "y": 0},

                {"x": 1, "y": 1},
                {"x": 2, "y": 1},
                {"x": 3, "y": 1},

                {"x": 1, "y": 2},
                {"x": 2, "y": 2},
                {"x": 3, "y": 2},
                {"x": 4, "y": 2},

                {"x": 1, "y": 3},
                {"x": 2, "y": 3},
                {"x": 3, "y": 3},

                {"x": 0, "y": 4},
                {"x": 1, "y": 4},
                {"x": 2, "y": 4},
                {"x": 3, "y": 4},
                {"x": 4, "y": 4}
            ]
        }
    }
}
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
This should fix the info.json lint.

File: keyboards/doodboard/duckboard/info.json [hunk 10]
```diff
@@ -0,0 +1,36 @@
+{
+    "keyboard_name": "duckboard",
+    "url": "",
+    "maintainer": "doodboard",
+    "width": 5,
+    "height": 5,
+    "layouts": {
+        "LAYOUT_default": {
```
Comment (fauxpark) [inline keyboards/doodboard/duckboard/info.json:8]:
```suggestion
        "LAYOUT": {
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/doodboard/duckboard/duckboard.h [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#pragma once
+
```
Comment (fauxpark) [inline keyboards/doodboard/duckboard/duckboard.h:19]:
```suggestion

#include "quantum.h"

```
And this should fix Travis.

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/doodboard/duckboard/readme.md [hunk 12]
```diff
@@ -0,0 +1,17 @@
+# duckboard
+
+An 18-key numpad with rotary encoder and OLED support. 
+
+* Keyboard Maintainer: [doodboard](https://github.com/doodboard)
+* Hardware Supported: duckboard PCB
+* Hardware Availability: [private group buy]
```
Comment (noroadsleft) [inline keyboards/doodboard/duckboard/readme.md:7]:
```suggestion
* Hardware Availability: Private Group Buy
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Comment by doodboard** [comment]
Thank yo I guys al life the reviews! Is there something else I need to do at this point?

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
> Thank you guys for the reviews! Is there something else I need to do at this point?

Nope; looks good, thanks!

**Pull request merged**